agent.py

```diff
@@ -347,16 +361,63 @@ def steps_queued(self):
         self.rewards = []
 
     @staticmethod
-    def unit_matrix(state, hero_unit, team_id, unit_types, only_self=False, max_units=16):
+    def unit_separation(state, team_id):
+        # Break apart the full unit-list into specific categories for allied and
+        # enemy unit groups of various types so we don't have to repeatedly iterate
+        # the full unit-list again.
+        allied_heroes       = []
+        allied_nonheroes    = []
+        allied_creep        = []
+        allied_towers       = []
+        enemy_heroes        = []
+        enemy_nonheroes     = []
+        enemy_creep         = []
+        enemy_towers        = []
+        for unit in state.units:
+            # FIXME - HACK - remove units that are invulnerable from consideration
+            # This fixes agents trying to attack towers they cannot
+            if is_invulnerable(unit):
```

Comment:
They still need to see the stats of a tower though. invulternability is just another scalar datapoint to a unit.